File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number):

    factors = set()
    result=""
    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")
    elif number == 1:
        factors.add(1)
        result="deficient"
    else:
        for i in range (1,number):
            if i < number/2 + 1 and number % i == 0:
                factors.add(int(i))
                if number/i < number:
                    factors.add(int(number/i))
    
        sum=0
        for i in factors:
             sum=sum+i
        
        if sum==number:
            result="perfect"
        elif sum > number:
            result="abundant"        
        else:
            result="deficient"
    return result
    

File: solutions/cpp/perfect-numbers/1/perfect_numbers.cpp
#include "perfect_numbers.h"

namespace perfect_numbers {
    classification classify(int number) {
        if ( number <= 0 ) {
                throw std::domain_error("");
        }
        int sum=1;
        int quotient=number/2;
        int divider=2;
        while (divider < quotient) {
            if (number % divider == 0) {
                quotient = number / divider;
                if (quotient != divider)
                    sum += quotient;
                sum += divider;
            }
            divider++;
        }
        if (number < 5)
            return classification::deficient;
        else if (number > sum)
            return classification::deficient;
        else if (number == sum)
            return classification::perfect;
        else
            return classification::abundant;
    }
} 